core/src/main/java/org/jobrunr/jobs/states/CarbonAwareAwaitingState.java

```diff
@@ -0,0 +1,99 @@
+package org.jobrunr.jobs.states;
+
+import org.jobrunr.jobs.Job;
+import org.jobrunr.scheduling.carbonaware.CarbonAwarePeriod;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Instant;
+
+import static java.lang.String.format;
+import static java.time.Duration.ofHours;
+import static java.time.Instant.now;
+import static java.util.Objects.isNull;
+
+public class CarbonAwareAwaitingState extends AbstractJobState {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CarbonAwareAwaitingState.class);
+
+    public static final int MINIMUM_CARBON_AWARE_SCHEDULE_INTERVAL_DURATION = 3;
+
+    private final Instant preferredInstant;
+    private final Instant from;
+    private final Instant to;
+
+    protected CarbonAwareAwaitingState() { // for json deserialization
+        super(StateName.AWAITING);
+        this.preferredInstant = null;
+        this.from = null;
+        this.to = null;
+    }
+
+    public CarbonAwareAwaitingState(CarbonAwarePeriod carbonAwarePeriod) {
+        this(carbonAwarePeriod.getFrom(), carbonAwarePeriod.getTo());
+    }
+
+    public CarbonAwareAwaitingState(Instant to) {
+        this(now(), to);
+    }
+
+    public CarbonAwareAwaitingState(Instant from, Instant to) {
+        this(null, from, to);
+    }
+
+    public CarbonAwareAwaitingState(Instant preferredInstant, Instant from, Instant to) {
+        super(StateName.AWAITING);
+        this.preferredInstant = preferredInstant;
+        this.from = from;
+        this.to = to;
+        validateCarbonAwarePeriod(from, to);
+    }
+
+    public Instant getPreferredInstant() {
+        return preferredInstant;
+    }
+
+    public Instant getFrom() {
+        return from;
+    }
+
+    public Instant getTo() {
+        return to;
+    }
+
+    public CarbonAwarePeriod getPeriod() {
+        return CarbonAwarePeriod.between(from, to);
+    }
+
+    public void moveToNextState(Job job, Instant idealMoment) {
+        moveToNextState(job, idealMoment, "Job scheduled at " + idealMoment + " to minimize carbon impact.");
+    }
+
+    public void moveToNextState(Job job, Instant idealMoment, String reason) {
+        if (!(job.getJobState() instanceof CarbonAwareAwaitingState)) {
+            throw new IllegalStateException("Only jobs in CarbonAwaitingState can move to a next state");
+        }
+        job.scheduleAt(idealMoment, reason);
+    }
+
+    public static void validateCarbonAwarePeriod(Instant from, Instant to) {
+        if (isNull(from) || isNull(to)) {
+            throw new IllegalArgumentException(format("'from' (=%s) and 'to' (=%s) must be non-null", from, to));
+        }
+        if (from.isAfter(to)) {
+            throw new IllegalArgumentException(format("'from' (=%s) must be before 'to' (=%s)", from, to));
+        }
+        // TODO isn't this dangerous for recurring jobs?
```

Comment:
I believe this was initially added to prevent users from making a mistake while scheduling jobs. We fail fast so they can check again that they really want to schedule carbon aware, and double check the margins.

If we no longer throw an exception, should we simply schedule the job as carbon aware?  Currently,
- if they set a deadline in the past, the job is immediately scheduled, and the reason of scheduling tells that the deadline is past.
- if they set a deadline lower than the allowed interval, the job is immediately scheduled. But the reason will depend on the availability of the forecast.

Reply:
> if they set a deadline in the past, the job is immediately scheduled, and the reason of scheduling tells that the deadline is past.
if they set a deadline lower than the allowed interval, the job is immediately scheduled. But the reason will depend on the availability of the forecast.

I like this approach. I also thought it was good to fail fast but I think in the long run it will cause us troubles.